Update specification of DNSSEC10 to handle nonstandard NSEC responses (#1478)
dotse/zonemaster: docs/public/specifications/tests/DNSSEC-TP/dnssec10.md

section 3.1.3, and the description of the NSEC RR itself is in
[RFC 4034][RFC 4034#section-4], section 4.

The specification of NSEC has been updated in [RFC 4470] and [RFC 9824], though
neither of those updates are directly relevant for this test case, some
implementations of [RFC 4470] produces an incorrect response of NSEC queries. The
response is a NODATA response with the NSEC record in the authority section
instead give a response with the NSEC record in the answer section. That

> deviation is handled in the test precedure.

Reviewer: ```suggestion
Note that the specification of NSEC has been updated in [RFC 4470] and [RFC 9824], and
although neither of those updates are directly relevant for this test case, some
implementations of [RFC 4470] produce an incorrect response to NSEC queries. The
response is a NODATA response with the NSEC record in the authority section
rather than in the answer section. That deviation is handled in the test precedure.
```

Author: I disagree that RFC 4470 and RFC 9824 are directly relevant for the test case. 

RFC 4470 is how the NSEC record is generated, but is does not change any properties that the test case tests. RFC 9824 introduces a psuedo RRtype, NXNAME, but if the client does not ask for it it must not be included.

Reviewer: I've just slightly reworded your original paragraph, I didn't change the meaning. See the diff.

Author: Your text is also fine.

| DS10_NSEC_RRSIG_EXPIRED            | ERROR   | ns_list, keytag                       | The RRSIG (signature) with tag {keytag} for the NSEC record has expired. Fetched from name servers "{ns_list}".                                                                             |
| DS10_NSEC_RRSIG_NOT_YET_VALID      | ERROR   | ns_list, keytag                       | The RRSIG (signature) with tag {keytag} for the NSEC record it not yet valid. Fetched from name servers "{ns_list}".                                                                        |
| DS10_NSEC_RRSIG_NO_DNSKEY          | WARNING | ns_list, keytag                       | There is no DNSKEY record matching the RRSIG (signature) with tag {keytag} for the NSEC record. Fetched from name servers "{ns_list}".                                                      |
| DS10_NSEC_RRSIG_VERIFY_ERROR       | ERROR   | ns_list, keytag                       | The RRSIG (signature) with tag {keytag} for the NSEC record cannot be verified. Fetched from name servers "{ns_list}".                                                                      |

> | DS10_NONSTANDARD_NSEC_RESPONSE     | NOTICE  | ns_list                               | The following name servers give a non-standard response to the NSEC query (unexpected NODATA response). Fetched from name servers "{ns_list}".                                              |

Reviewer 2: Maybe give a little bit more detail on what’s wrong?
```suggestion
| DS10_NONSTANDARD_NSEC_RESPONSE     | NOTICE  | ns_list                               | The following name servers give a non-standard response to the NSEC query (NSEC RR in authority section instead of answer section). Fetched from name servers "{ns_list}".       |
```

             2. If the number of NSEC records is greater than one then add name
                server IP to the *Erroneous Multiple NSEC* set.
             3. Else, if the owner name of the NSEC record is not *Child Zone*
                then add name server IP to the *NSEC Mismatches Apex* set.
          2. Else add the name server IP to the
             *NSEC Query Gives Erroneous Answer* set.
       3. Else if the answer section is empty and the authority section contains
          an NSEC record then do:
          1. Add the name server IP to the *NSEC Query gives NSEC In Response* set and the
             *NSEC Query Gives Nonstandard NSEC NODATA* set.
          2. If the SOA record is missing from the authority section then add name
             server IP to the *NSEC NODATA Missing SOA* set.
          3. Else if the owner name of SOA record is not *Child Zone* then
             add name server IP and owner name to the *NSEC NODATA Wrong SOA*
             set.
          4. If the authority section contains more than one NSEC record then
             add name server IP to the *Erroneous Multiple NSEC* set.

>           5. Else do: #

Reviewer 2: ```suggestion
          5. Else do:
```

             1. If the owner name of the NSEC record is not *Child Zone* then
                add name server IP to the *NSEC Mismatches Apex* set.
             2. Else if the type list in the NSEC record matches at least one
                of the following criteria then add name server IP to the
                *NSEC Incorrect Type List* set:

>                 1. At least one of SOA, NS, DNSKEY, NSEC or RRSIG is missing.
>                 2. At least one of NSEC3PARAM or NSEC3 is included.

Reviewer: Looking at the domains/name servers in #1424, it seems that the covered types in the NSEC RR include NSEC3PARAM. Should the type list validation be updated or just skipped in this case?
```
$ dig +dnssec NSEC uspory.cz @205.251.197.186

; <<>> DiG 9.18.33-1~deb12u2-Debian <<>> +dnssec NSEC uspory.cz @205.251.197.186
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 50136
;; flags: qr aa rd ad; QUERY: 1, ANSWER: 0, AUTHORITY: 4, ADDITIONAL: 1
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags: do; udp: 4096
;; QUESTION SECTION:
;uspory.cz.                     IN      NSEC

;; AUTHORITY SECTION:
uspory.cz.              900     IN      SOA     ns-1466.awsdns-55.org. awsdns-hostmaster.amazon.com. 1 7200 900 1209600 86400
uspory.cz.              900     IN      RRSIG   SOA 13 2 900 20260428165716 20260428144216 14144 uspory.cz. eu+g4dJ4MDTANJCqGJ5+uqPGecsouMOWr/4iNYUa35qfnYplNeWwu39e bxe/WyIVyetopxYCBBACFyiv6tZjgg==
uspory.cz.              86400   IN      NSEC    \000.uspory.cz. A NS SOA PTR HINFO MX TXT RP AAAA SRV NAPTR SSHFP RRSIG DNSKEY NSEC3PARAM TLSA CDS CDNSKEY SVCB HTTPS SPF CAA
uspory.cz.              86400   IN      RRSIG   NSEC 13 2 86400 20260429164216 20260428144216 14144 uspory.cz. Dabc0d/UWK1s/FhBERJGKIVqBDhmt4cm5tPhwuLsO0voKC9uBm/nIW98 jX1yqJMh68a8hc1IAqfxN1N7qotm7Q==

;; Query time: 3 msec
;; SERVER: 205.251.197.186#53(205.251.197.186) (UDP)
;; WHEN: Tue Apr 28 17:42:16 CEST 2026
;; MSG SIZE  rcvd: 37
```

```suggestion
                1. At least one of SOA, NS, DNSKEY, NSEC or RRSIG is missing.
                2. At least NSEC3 is included.
```

Author: I strongly disagree with the proposed change. It is violating the standards to annouce NSEC3PARAM records in NSEC zones. We should keep it as is.

Reviewer: "Expected" behavior from Route53 implementation, see https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/dns-configuring-dnssec-proof-of-nonexistence.html:

> In cases where there is a record matching the query name but not the query type (such as querying for web.example.com/AAAA but there is only web.example.com/A present), we return a minimal NSEC (next secure) record containing all supported resource record types.

But that is yet another peculiarity to handle, as currently Zonemaster will output `DS10_NSEC_ERR_TYPE_LIST` ERROR messages.

Author: We could lower the level to NOTICE which is probably more reasonable.

Author: Maybe expected from Route53, but not from the DNSSEC standards.

Reviewer: > Maybe expected from Route53, but not from the DNSSEC standards.

Yes but if we didn't do anything for that too then we'd be having the same concerns from the same users.

> We could lower the level to NOTICE which is probably more reasonable.

Yes great !
